app/frontend/wallet/helpers/CachedDeriveXpubFactory.ts

```diff
@@ -16,9 +21,13 @@ function CachedDeriveXpubFactory(derivationScheme, deriveXpubHardenedFn) {
       * since a failure (e.g. rejection by user) there leads to
       * the creation of a fresh wallet and crypto provider instance
       */
-      derivedXpubs[memoKey] = deriveHardened
-        ? deriveXpubHardenedFn(absDerivationPath)
-        : deriveXpubNonhardenedFn(absDerivationPath)
+
+      if (deriveHardened) {
+        const pubKeys = await deriveXpubHardenedFn(absDerivationPath)
+        Object.assign(derivedXpubs, pubKeys)
+      } else {
+        derivedXpubs[memoKey] = deriveXpubNonhardenedFn(absDerivationPath)
+      }
     }
 
     /*
```

Comment:
if we are awaiting on the pubKeys bulk object before assigning it to derivedXpubs, then this comment is no longer true and should be removed. To be fair though - concurrent export of public keys should not be happening anywhere and even if it by any chance did, it would result in a duplicate prompt which would be annoying but not fatal

Reply:
But we still store them as promises, all of them, so currently it is still true. And afaik, concurrent export of public keys is currently happening and was until now, (for example addresses have the same parent path) and these are awaiting the same promise.

Reply:
yes, but if you do

```
const pubKeys = await deriveXpubHardenedFn(absDerivationPath)
Object.assign(derivedXpubs, pubKeys)
```
That defeats the point of storing the derived keys as promises because if somebody executes asynchronously the function to derive a key, the promise won't be in the derivedXpubs dictionary yet